### ISSUE TITLE
Add $transaction_id to order->payment_complete() filters/actions

### ISSUE THREAD

**Issue by TheRealSimpleSimon** [woocommerce/woocommerce#28874]
There is no way for a filter/action to see what the incoming $transaction_id value is.
Location: woocommerce/includes/class-wc-order.php line 99
public function payment_complete( $transaction_id = '' ) ....

In some cases, this can be important to determine the proper actions to take.
For example, it may be desired to not set date_paid if there is no transaction_id.
(line 111: filter woocommerce_valid_order_statuses_for_payment_complete)

Also, the desired new status may depend on the incoming $transaction_id.
(line 118: filter woocommerce_payment_complete_order_status)

For completeness, the following probably should also be enhanced:
(line 105: action woocommerce_pre_payment_complete)
(line 121: action woocommerce_payment_complete)
(line 123: action woocommerce_payment_complete_order_status_ . $this->get_status())
* Possibly the above two actions should also have "$this" added to the paramater list, also.

I am currently exploring the use of woocommerce_before_order_object_save as a workaround,
to trap out undesired changes created by payment_complete.

However, using that filter introduces unnecessary overhead because it will be called much more
often than is needed to achieve the desired result.

**Comment by roykho** [comment]
Hmmm since you get the order id from the action then why not:
```
$order = wc_get_order( $order_id );
$transaction_id = $order->get_transaction_id();
```

**Comment by TheRealSimpleSimon** [comment]
> Hmmm since you get the order id from the action then why not:
> 
> ```
> $order = wc_get_order( $order_id );
> $transaction_id = $order->get_transaction_id();
> ```

Hi!
Thanks for reading my issue!

Well, my suggestion to add the current object to those two actions is not very important - but as the object is already referenceable, it can save overhead by just passing it (as is done for the filters 'nearby').

get_transaction_id() won't work because what I consider the 'right' place to do the work is before the transaction_id value (if any) has been added to the order. Take a look at the function in question (starts at class-wc_order.php:99) and you'll see. :)

**Comment by roykho** [comment]
What about `woocommerce_pre_payment_complete` hook? That is before the transaction id is set.

**Comment by TheRealSimpleSimon** [comment]
> What about `woocommerce_pre_payment_complete` hook? That is before the transaction id is set.

Oh - Thank you! I did forget to mention that one at line 105, I just edited in to the original post.

That one could be helpful - but it does not give access to the incoming $transaction_id value) either.

Also, in some situations you may want to simply control what happens in the 'if' at line 111 -
although arguments can be made that returning a value to force the if to be false is improper usage.
My rebuttal to that is along the lines of "no transaction_id means there is no "valid status for payment complete" -
which is in fact, the exact situation I am trying to address with this issue.

$transaction_id should be made available to all 5 hooks (2 filters, 3 actions)

**Comment by roykho** [comment]
> but it does not give access to the incoming $transaction_id value either.

True but you can still get it as I described previously.

> Also, in some situations you may want to simply control what happens in the 'if' at line 111

That's only checking valid order statuses. If you have custom statuses you want to check against you can do that via the filter and `woocommerce_valid_order_statuses_for_payment_complete` has the order object passed.

**Comment by TheRealSimpleSimon** [comment]
> > but it does not give access to the incoming $transaction_id value either.
> 
> True but you can still get it as I described previously.
> 
> > Also, in some situations you may want to simply control what happens in the 'if' at line 111
> 
> That's only checking valid order statuses. If you have custom statuses you want to check against you can do that via the filter and `woocommerce_valid_order_statuses_for_payment_complete` has the order object passed.

Again, at the time in question, the incoming $transaction_id exists nowhere but in the current run context of the payment_complete function - where it is passed as the one (optional) parameter to the function. The earliest it could be accessed from anywhere else is after line 113: $this->set_transaction_id( $transaction_id );

Note that if at line 111 controls the set of transaction_id and date_paid - which is why I mentioned that some would consider that an improper usage. But the woo code is doing more than just checking status - it is determining if the current order status is "valid for payment complete", and if so, sets the two meta values -- and it also might change the current order_status (line 118).
These 3 things are exactly what I want to stop. Aside, I do have several custom statuses, and would not want them to be modified as they don't have anything to do with whether an order has been paid for or not.

This is why the workaround I am exploring (woocommerce_before_order_object_save) should make it possible to back-out the undesired changes before they actually hit the database.

In other words, my client's requirement is to separate fulfillment from payment - which is typical in B2B scenarios.
I have a few other customizations in place to make this work - and it's all perfect -- except for this one thing.
To oversimplify, the client doesn't want the order edit page to show "date paid xxxx" when it hasn't actually been paid.

**Comment by TheRealSimpleSimon** [comment]
Oh - if the core code would look at the incoming transaction_id, and just quit if it's null, that would work, too.

Therefore, calling a brand-new filter to help decide that would be another decent solution -
but more complicated to implement.

**Comment by TheRealSimpleSimon** [comment]
Further research:
I see no need to change 'woocommerce_payment_complete'.
It is only invoked after the set_transaction_id function has been called, so if anyone needs it, it's available.
In fact, I may be able to trap out the "bad" (from my point of view) updates using this hook, and "un-payment-complete" the order here instead of via 'woocommerce_before_order_object_save'. Further research is needed.

The call to 'woocommerce_payment_complete_order_status' at includes/class-wc-order.php line 291 (in 'maybe_set_date_paid') can stand as-is. If the target function of the filter cares about $transaction_id, it should define it as 'optional'.
The other call to 'woocommerce_payment_complete_order_status' is at line 834 (in function get_date_paid) and can also stand as-is for the same reasons.
I believe that any customization using this filter that cares about $transaction_id will just return array('completed') from this filter if the transaction_id is missing from both the parameter list and the order's metadata, and it will all work as desired (both truth table and state/event analysis seem to be OK).

Again, the point of all this is to change the definition of 'paid' from "date_paid' being set to 'transaction_id' being set.
All of the other necessary hooks (eg. 'woocommerce_order_is_paid') to accomplish this exist -
this issue (no access to $transaction_id) is the only roadblock.

Footnote: if the B2B environment that wants to use this type of customization is also using stock management, the site developer will need to research includes/wc-stock-functions.php depending on how they want "invoiced orders" to work for them in conjunction with how they manage their packing and shipping operations.

**Comment by davefx** [comment]
@TheRealSimpleSimon @roykho 
I've prepared PR #28930 with the needed changes for this. I've found myself in the need of having this extra parameter to be able to fill in the transaction_id in some cases in which the payment is complete when the order has already been moved to processing status (due to the behaviour of some third-party plugins).

**Comment by TheRealSimpleSimon** [comment]
> @TheRealSimpleSimon @roykho
> I've prepared PR #28930 with the needed changes for this. I've found myself in the need of having this extra parameter to be able to fill in the transaction_id in some cases in which the payment is complete when the order has already been moved to processing status (due to the behaviour of some third-party plugins).

Hi - looks good (if I'm reading that stuff correctly) - except you did not update the filters - both of which need the transaction_id in order to make their processing decision at the point it is needed.

For example, at line 111, having the transaction_id value (or more commonly, "no value") allows the filter to return a "non-existent" status, which would in turn make the has_status call return false, and therefore prevent the set_date_paid at line 116 from executing. This allows the user customization to be a clean 3-4 lines instead of a couple of dozen or more lines to revert the date_paid. For example, the filter could look like:
    if ( is_empty($transaction_id) { return array('no-no-no');  }
Similar applies to the set_status at line 118. 

The business rule that needs the enhancement to woocommerce_payment_complete_order_status is:
  "No transaction_id means the order is not paid - even though some code somewhere thinks it is".
This is my situation.

Having the transaction_id in filter woocommerce_payment_complete_order_status at line 118 would provide for setting a custom order status of, for example, "make sure we got paid for this". I personally do not need this one (because I want to stop 112-121 from executing at all if there's no transaction_id value), but for completeness and other possible applications, it should be there.

**Comment by davefx** [comment]
@TheRealSimpleSimon 
Thanks for your review.
I just added the extra parameter to the two remaining filters.
